### PR TITLE
[WIP] Added new egs for dsb2018 nuclei detection

### PR DESCRIPTION
**Pull request by YiwenShaoStephen**

### CODE REVIEW

File: egs/dsb2018/v1/local/training.py [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+parser.add_argument('--name', default='Unet-5', type=str,
+                    help='name of experiment')
+parser.add_argument('--train-dir', default='data/train_val/split0.9_seed0', type=str,
+                    help='Directory of processed training and validation data')
+parser.add_argument('--test-dir', default='data/test', type=str,
+                    help='Directory of processed test data')
+parser.add_argument('--num-classes', default=2, type=int,
+                    help='Number of classes to classify')
+parser.add_argument('--tensorboard',
+                    help='Log progress to TensorBoard', action='store_false')
```
Comment (hhadian) [inline egs/dsb2018/v1/local/training.py]: Do we need tensorboard? 
I tried to run your code previously (from your repo, not this PR) but I couldn't install tensorboard_logger so I disabled it (and I ran it w/o issues).
I guess it's good to minimize the requirements.

File: egs/dsb2018/v1/run.sh [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+depth=5
+epochs=10
+height=128
+width=128
+batch=16
+
+
+. ../../../scripts/parse_options.sh  # e.g. this parses the --stage option if supplied.
+
+pip install -r local/requirements.txt
```
Comment (hhadian) [inline egs/dsb2018/v1/run.sh]: shouldn't this be pip3? (also I guess you need to add `--user`)
BTW, are you using miniconda in this PR? It seems that you are not.

File: egs/dsb2018/v1/run.sh [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+if [ $stage -le 0 ]; then
+  # data preparation
+  local/prepare_data.sh --train_prop $train_prop --seed $seed
+fi
+
+name=Unet_${depth}_${epochs}_sgd
+if [ $stage -le 1 ]; then
+  # training the network
+  CUDA_VISIBLE_DEVICES=$gpu_id ./local/training.py \
```
Comment (hhadian) [inline egs/dsb2018/v1/run.sh]: We should think of some way to make this safe in regards to CLSP gpu reservation. 
I guess we could have a cmd.sh like in Kaldi and use queue.pl and run.pl. 
At least for this task (which is very small), we can simply use `queue.pl --gpu 1` to run it in one stage.
@danpovey and @jtrmal, what would you suggest?

File: egs/dsb2018/v1/run.sh [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+. ../../../scripts/parse_options.sh  # e.g. this parses the --stage option if supplied.
+
+pip install -r local/requirements.txt
+
+if [ $stage -le 0 ]; then
+  # data preparation
+  local/prepare_data.sh --train_prop $train_prop --seed $seed
+fi
+
+name=Unet_${depth}_${epochs}_sgd
```
Comment (danpovey) [inline egs/dsb2018/v1/run.sh]: please don't use uppercase in directory names.

File: egs/dsb2018/v1/local/training.py (hunk 1 above)
Comment (YiwenShaoStephen) [reply egs/dsb2018/v1/local/training.py]: We don't necessarily have to install it. If enabled, it will write tensorboard events that can help us analyse the training process. Otherwise, we can only get information from std:out

File: egs/dsb2018/v1/run.sh (hunk 2 above)
Comment (YiwenShaoStephen) [reply egs/dsb2018/v1/run.sh]: I assume we will add miniconda3 to the python path as it suggests. And in this case, pip will point to the pip inside miniconda3 so we don't need --user option here.

File: egs/dsb2018/v1/run.sh (hunk 4 above)
Comment (YiwenShaoStephen) [reply egs/dsb2018/v1/run.sh]: OK, I will change it to lowercase.

File: egs/dsb2018/v1/run.sh (hunk 3 above)
Comment (YiwenShaoStephen) [reply egs/dsb2018/v1/run.sh]: OK, I will do it in this way.

File: egs/dsb2018/v1/local/process_data.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    args = parser.parse_args()
+
+    train_output = "{0}/train_val/split{1}_seed{2}/train.pth.tar".format(
+        args.outdir, args.train_prop, args.seed)
+    val_output = "{0}/train_val/split{1}_seed{2}/val.pth.tar".format(
+        args.outdir, args.train_prop, args.seed)
+    if not (os.path.exists(train_output) and
+            os.path.exists(val_output)):
+        random.seed(args.seed)
+        train, val = DataProcess(args.train_input,
```
Comment (danpovey) [inline egs/dsb2018/v1/local/process_data.py]: Just curious: is this how data preprocessing is normally done in pytorch?
I'm concerned that this saves it all in memory at once, so would not scale that well.
When the training code loads this data, does it also store it all in memory?

File: egs/dsb2018/v1/local/dataset.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        img = data['img'].numpy()
+        height, width, channel = img.shape
+        img = self.transformation(img)
+
+        # bounding box
+        num_offsets = len(self.offset_list)
+        mask = data['mask'].numpy()
+        bound = torch.zeros(num_offsets, self.height, self.width)
+
+        for k in range(num_offsets):
```
Comment (danpovey) [inline egs/dsb2018/v1/local/dataset.py]: I think we need some explanation in comments about what is going on here.  This is not obvious at all.

File: egs/dsb2018/v1/local/training.py [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+    args = parser.parse_args()
+
+    if args.tensorboard:
+        from tensorboard_logger import configure
+        print("Using tensorboard")
+        configure("exp/%s" % (args.name))
+
+    s_trans = tsf.Compose([
+        tsf.ToPILImage(),
+        tsf.Resize((args.img_height, args.img_width)),
```
Comment (danpovey) [inline egs/dsb2018/v1/local/training.py]: I wonder when this gets applied to the supervision, does it attempt to interpolate the 'mask' values which represent integer object-ids?

File: egs/dsb2018/v1/local/training.py [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+                       args.num_classes, args.img_height, args.img_width)
+    trainloader = torch.utils.data.DataLoader(
+        trainset, num_workers=1, batch_size=args.batch_size)
+
+    valset = Dataset(val_data, s_trans, offset_list,
+                     args.num_classes, args.img_height, args.img_width)
+    valloader = torch.utils.data.DataLoader(
+        valset, num_workers=1, batch_size=args.batch_size)
+
+    NUM_TRAIN = len(trainset)
```
Comment (hhadian) [inline egs/dsb2018/v1/local/training.py]: better to change these to lower-case, or maybe just use len(...) in the print() call

File: egs/dsb2018/v1/local/training.py (hunk 7 above)
Comment (YiwenShaoStephen) [reply egs/dsb2018/v1/local/training.py]: The transformation is applied on the offset feature maps instead of the 'maks' values that represent object-ids.
The offset feature maps are 'bound' in my code, where the value is either 1 or 0 that represents whether pixel (x, y) is of the same object as its offset pixel (x+i, y+j). And the interpolate also applies to this bound matrix, so that after transformations, there will appear value between 0 and 1.
